src/Moq/Protected/ProtectedMock.cs

```diff
@@ -510,16 +499,57 @@ private static Type[] ToArgTypes(object[] args)
 			return types;
 		}
 
+		private static FieldInfo ItRefAnyField(Expression expr)
+		{
+			FieldInfo itRefAnyField = null;
+
+			if (expr.NodeType == ExpressionType.MemberAccess)
+			{
+				var member = (MemberExpression)expr;
+				if (member.Member is FieldInfo field)
+				{
+					if (field.Name == nameof(It.Ref<object>.IsAny))
+					{
+						var fieldDeclaringType = field.DeclaringType;
+						if (fieldDeclaringType.IsGenericType)
+						{
+							var fieldDeclaringTypeDefinition = fieldDeclaringType.GetGenericTypeDefinition();
+							if (fieldDeclaringTypeDefinition == typeof(It.Ref<>))
+							{
+								itRefAnyField = field;
+							}
+						}
+					}
+				}
+			}
+
+			return itRefAnyField;
+		}
+
 		private static Expression ToExpressionArg(Type type, object arg)
 		{
-			if (arg is LambdaExpression lambda)
+			if (arg is LambdaExpression lambda && !typeof(LambdaExpression).IsAssignableFrom(type))
 			{
 				return lambda.Body;
 			}
 
 			if (arg is Expression expression)
 			{
-				return expression;
+				if (!typeof(Expression).IsAssignableFrom(type))
+				{
+					return expression;
+				}
+
+				if (expression.IsMatch(out _))
+				{
+					return expression;
+				}
+				
+				if (ItRefAnyField(expression) != null)
+				{
+					return expression;
+				}
```

Comment:
You're likely more up-to-date about this code than I am, so perhaps you could quickly explain the following points to me?:

* Is there a reason why the first check for `LambdaExpression`s precedes / is outside the `if (arg is Expression expression)` block? It would seem more logical to move it in there with the other `Expression` tests, so that at the top level, we'd only have a single block that deals with all special cases involving `Expression`.

* `!typeof(Expression).IsAssignableFrom(type)` &ndash; this appears to be the test to see whether (`arg` aka) `expression` could serve as a valid value for a `type`-typed parameter. Something doesn't feel quite right about the `typeof(Expression)` constant... shouldn't we be more precise and use `expression`'s actual type, i.e. shouldn't the check be `type.IsAssignableFrom(expression.GetType())`?

Reply:
Agreed on both points.